gpii/node_modules/preferencesServer/src/preferencesService.js

```diff
@@ -73,12 +125,231 @@ var fluid = fluid || require("infusion");
+                args: ["{arguments}.0"],
+                namespace: "findById"
+            }, {
+                func: "gpii.dbOperation.verifyExists",
+                args: ["{arguments}.0", "prefsSafe", "The prefs safe does not exist"]
+            }, {
+                func: "{expressUserUtils}.createNewUser",
+                args: [{username: "{arguments}.1.username", password: "{arguments}.1.password"}],
+                namespace: "createNewUser"
+            }, {
+                funcName: "gpii.dbOperation.dbDataStore.addRecord",
+                args: ["{dbDataStore}.saveDataSource",
+                       "gpiiCloudSafeCredential",
+                       "id", {
+                           type: "gpiiCloudSafeCredential",
+                           schemaVersion: "0.1",
+                           prefsSafeId: "{arguments}.1.prefsSafeId",
+                           gpiiExpressUserId: "{arguments}.0._id"
+                       }],
+                namespace: "addRecord"
+            }, {
+                funcName: "{dataStore}.findById",
+                args: ["{arguments}.0.id"]
+            }],
+            onCloudSafeUnlock: [{
+                func: "{expressUserUtils}.unlockUser",
+                args: ["{arguments}.0.username", "{arguments}.0.password"], // username, password
+                namespace: "unlockUser"
+            }, {
+                func: "{dataStore}.findSafeByExpressUserLookup",
+                args: ["{arguments}.0._id"],
+                namespace: "findSafeByExpressUserLookup"
+            }, {
+                func: "{dataStore}.findById",
+                args: ["{arguments}.0.prefsSafeId"],
+                namespace: "findById"
             }]
         }
     });
 
     // APIs for Preferences and Preferences Safes
 
+    gpii.preferencesServer.preferencesService.errors = fluid.freezeRecursive({
+        gpiiKeyMissing: "GPII key \"%gpiiKey\" does not exist",
+        gpiiKeyExisted: "GPII key \"%gpiiKey\" already exists",
+        noUpdateOnSnapset: "Cannot update:  GPII key \"%gpiiKey\" is a snapset"
+    });
+
+    /**
+     * Returns a preferences safe using it's id. If the safe is not found an error object
+     * is returned.
+     *
+     * @param  {Object} dataStore - Instance of `gpii.dbOperation.dataStore`
+     * @param  {String} prefsSafeId Prefs Safe ID
+     * @return {Promise}            Returns a promise resolving with the preferences safe,
+     * or rejecting with an error payload containing an error `message`.
+     */
+    gpii.preferencesServer.preferencesService.getPrefsSafe = function (dataStore, prefsSafeId) {
+        var promiseTogo = fluid.promise();
+        dataStore.findById(prefsSafeId).then(
+            function (data) {
+                if ($.isEmptyObject(data) || data.type !== "prefsSafe") {
+                    promiseTogo.reject({
+                        isError: true,
+                        message: "Cannot lookup prefsSafe",
```

Comment:
With recent changes to the preferences server, a collection of error messages with "errorCode" is defined at [gpii.preferencesServer.errors[]](https://github.com/GPII/universal/blob/master/gpii/node_modules/preferencesServer/src/preferencesServerConst.js#L26) and shared across preferencesServer module. Could you follow the same pattern for error handling?

BTW, it would be helpful to think about if some similar messages can be replaced by one common message. For example, "Cannot lookup prefsSafe" here and "Error fetching prefsSafe" at line 207 probably can share one of these?

Reply:
I've moved all of these into the errors, and am trying to reuse a few of them with when possible, although many require a specific check or comment that is unique.